### PR TITLE
Editor: Reset Post Edits when switching posts in the post editor

### PR DESCRIPTION
**Pull request by youknowriad**
If we create a new post, switch to a draft from the sidebar without saving and coming back to /post/new again, some initial post edits were kept instead of having a fresh new post. Currently, we can experience this for the "sticky" post flag for example. 

While we continue reduxifying the editor, this behaviour will be more and more troublesome. (see #8814 )

This PR resets the postEdits while switching posts on the PostEditor Component.

**Testing instructions**

 - Start new post
 - Click on the "sticky" post icon
 - Before post saves, open drafts drawer and select another post
 - Accept warning that changes will be lost
 - Observe draft loads correctly
 - Start new post
 - The "sticky" post icon should not be selected, we should have a fresh post

### CODE REVIEW

File: client/post-editor/post-editor.jsx [hunk 1]
```diff
@@ -156,6 +156,12 @@ export const PostEditor = React.createClass( {
 
 	componentWillReceiveProps: function( nextProps ) {
 		const { siteId, postId } = this.props;
+
+		// When switching posts, reset post edits for the post we're leaving
+		if ( nextProps.postId !== postId ) {
```
Comment (aduth) [inline client/post-editor/post-editor.jsx]: A very edge case that might occur is switching from editing posts of the same ID on two different sites. Should we also check whether the site ID has changed?

Comment (youknowriad) [inline client/post-editor/post-editor.jsx]: Yeah, I think we should. I though these IDs were unique between sites.

**Comment by youknowriad** [comment]
@aduth The use-case you mentioned is probably worth fixing since the title/content is reset if we trigger a new Post from the master bar. Fixed in f38d7ba

File: client/post-editor/controller.js [hunk 2]
```diff
@@ -115,6 +115,7 @@ module.exports = {
 
 		function startEditing( siteId ) {
 			context.store.dispatch( setEditorPostId( postID ) );
```
Comment (aduth) [inline client/post-editor/controller.js]: Just a thought - should the reducer for `state.posts.edits` be listening for the `EDITOR_POST_ID_SET` action type and reset itself internally, rather than us calling the reset action explicitly? A possible complication from this is that we might not know that the ID is different, though ideally we're only dispatching this action in cases where the ID has in-fact changed. Or we could track ID in `edits` itself when `EDITOR_POST_ID_SET` is dispatched and compare against it in the reducer.

Comment (aduth) [inline client/post-editor/controller.js]: I agree that a more meaningfully named action creator might help avoid accidents. Do you think it would be used in addition to or in place of the existing `setEditorPostId` ? If in addition to, would we treat it as a totally separate action type or could we extend the existing action type with additional details / meta to describe it as being a "new" editing session?

>Also, does this mean, we should drop the current `resetPostEdits` calls (`componentWillUnmount`, `componentWillReceiveProps`, `onSaveSuccess`)? I would say yes, but maybe this would change the behavior a bit.

I think it'd be nice to allow the editor component to not worry about these details. If I were to guess, I suspect we reset upon unmounting as a sort of clean-up task, so the state doesn't hang around after leaving the editor. As long as we properly reset when revisiting the editor, I don't know that there will be any problems. Otherwise we could fire some new `stopEditing` action creator upon unmounting / leaving the editor that has similar effects of resetting edit state.

Comment (aduth) [inline client/post-editor/controller.js]: >Otherwise we could fire some new `stopEditing` action creator upon unmounting / leaving the editor that has similar effects of resetting edit state.

This is now ringing bells that we'd previously reset edits on a `page.exit` handler but changed away from that approach for some reason. Let me see if I can dig that up.

Comment (youknowriad) [inline client/post-editor/controller.js]: > Do you think it would be used in addition to or in place of the existing setEditorPostId ? 
I was thinking as a replacement on the controller, but I'm still wondering if I should keep the `setEditorPostId` for the specific usecase when saving a new post with ID yet, we'll have to update the ID without "starting a new editing session".

Let me try to put that in place, to have more context for discussion

Comment (youknowriad) [inline client/post-editor/controller.js]: Yes, so I added the `startPostEditor` action in 4c6278d but I kept the cleaning on unmount, because I was experiencing the same bug in #4539. SO any thoughts about the current implementation? Is this new Action worth adding? It looks ok, but the initial change was 4 lines of code :P

Comment (youknowriad) [inline client/post-editor/controller.js]: Thought, this seems more "logic"

Comment (youknowriad) [inline client/post-editor/controller.js]: Also just a small remark, I'm wondering why the lingering edits change the behavior of the "New" links. They seem unrelated to me. Maybe a bug on these links instead no?

Comment (aduth) [inline client/post-editor/controller.js]: >Also just a small remark, I'm wondering why the lingering edits change the behavior of the "New" links. They seem unrelated to me. Maybe a bug on these links instead no?

There's an expanded explanation for the behavior on #4539

Comment (youknowriad) [inline client/post-editor/controller.js]: I think what you proposed: `getNewEditorPath` selector, is a good approach to solve this definitely.

Anyway, Should we merge the current approach of this branch?

**Comment by aduth** [comment]
Brain-dumping here, since I'm concerned we have too many ways of changing the ID of the edited post and too many places where we're resetting edits. Taking a step back and trying to think through expectations: If we've changed posts or navigated away from the editor, it makes sense to drop unsaved edits for the post we had been editing. The main complication here is when a new draft is saved, since we transition from an empty post ID to the draft's new saved ID.

I would like to think in that case that it would be fine to reset edits for the post, but I worry that if a user continues to make edits while a save request is in flight, those edits need to be preserved until next we save, so we can't simply reset the edits upon the save completing.

With this in mind, perhaps it is best that the `setEditorPostId` doesn't have any effect on the edits state. That's where action creators triggered upon starting an editing session or leaving an editing session might be more appropriate, like your proposed `startPostEditor`.

But then, should those edits be reset when starting or ending? I mentioned we don't want lingering state hanging around after leaving the editor, but we similarly need to reset edits when changing between posts while still in the editor (drafts drawer, master bar "Write").

As an aside, we already have [some logic](https://github.com/Automattic/wp-calypso/blob/5c6b2875509542112588df88cd35680cbcadbb6b/client/post-editor/controller.js#L118) which needs to run when starting a post to assign the post's type, so this could fit well within an action creator which signals that we've started editing a post.

With these changes, what we have is resetting edits both when starting a post and when ending, but implemented in different ways. The resetting upon starting occurs in the controller by calling `startPostEditor` and when leaving in the [component's `componentWillUnmount` calling `resetPostEdits`](https://github.com/Automattic/wp-calypso/blob/5c6b2875509542112588df88cd35680cbcadbb6b/client/post-editor/post-editor.jsx#L143-L144).

Could we make this more consistent? Perhaps your solution in f38d7bac74fd15aecb473c09fa8d399f67f45ffd  is best in simply calling `resetPostEdits` in the controller explicitly.

Another idea is to have special-purpose `startEditingPost` and `stopEditingPost`, both in the controller, when entering and leaving the route respectively.

Or maybe we don't care if edits linger after leaving the editor, as long as we fix the bugs observed in #4539. That way we don't need to reset anything when leaving the editor.

If we do keep the new action creator, do you think it could simply extend `setEditorPostId`, something like:

```js
export function startEditingPost( siteId, postId ) {
	return {
		...setEditorPostId( postId ),
		clearEdits: true
	};
}
```

... since almost all of the behavior is the same except the reducer logic for resetting? There we could just check for the presence of this new key.

This was long, sorry 😄  Just expressing some inner turmoil being uncertain what's the clearest solution.

**Comment by youknowriad** [comment]
Yeah difficult problem to solve and I see two way of thinking here: 

1- Should we continue our Redux `setters` approach, which is simpler for this particular case (but maybe not for the longer term). In this case the solution would be:

 - Call `resetPostEdits` where needed: Unmounting and switching site while still on the editor like in f38d7ba And we could resolve the edgecase you mentioned about resetting edits for new posts being saved by checking that the old id should not be `null` before calling the reset combined with a state variable maybe (I have to look deeper to find the correct test to add)

2- The second solution is I think a more "redux" way of thinking but It requires bigger changes (and I'm not sure it's worth it). In this solution, we have to rethink our actions as a whole. 

 - Having `startEditingPost`, `stopEditingPost` both in the controller as you said.
 - Having a `savePost` which I assume already exists (and we can use to reset the edits as well). We need a way in this action to tell, that we're saved a new draft (without id), or we updated an existing post.

With this solution, all resets should be handled on the reducers, and the component should not worry about "resetting" the edits explicitly.

That said, I'm more keen to try the second.

**Comment by youknowriad** [comment]
@aduth I tried the new approach (pure redux actions) in bcfc667 and I'm pretty satisfied with the results. What do you think?

File: client/post-editor/post-editor.jsx [hunk 3]
```diff
@@ -695,19 +692,12 @@ export const PostEditor = React.createClass( {
 			this.props.resetEditorLastDraft();
 		}
 
-		// Assign editor post ID to saved value (especially important when
-		// transitioning from an unsaved post to a saved one)
-		if ( post.ID !== this.props.postId ) {
-			this.props.setEditorPostId( post.ID );
-		}
+		// Remove this when the editor is completely reduxified ( When using Redux actions for all post saving requests )
+		this.props.savePostSuccess( post.site_ID, this.props.postId, post, {} );
```
Comment (aduth) [inline client/post-editor/post-editor.jsx]: Only by a bit of luck does this not cause a notice to be shown (which would be non-ideal, at least for now, given that it's redundant with the editor's own notice). See: https://github.com/Automattic/wp-calypso/blob/795cd4d6b863be46f567d689f81c0538a07c6074/client/state/notices/middleware.js#L78-L93 (it would show a notice if `{}` instead included a status property).

Arguably this is an argument against the lack of discoverability of middlewares 😄

File: client/state/posts/reducer.js [hunk 4]
```diff
@@ -261,7 +263,16 @@ export function edits( state = {}, action ) {
...
+		case EDITOR_START:
+			return Object.assign( {}, state, {
+				[ action.siteId ]: {
+					...state[ action.siteId ],
+					[ action.postId || '' ]: { type: action.postType }
+				}
+			} );
+
+		case EDITOR_STOP:
+		case POST_SAVE_SUCCESS:
```
Comment (aduth) [inline client/state/posts/reducer.js]: I wondered if clearing edits on success would have a negative impact where we currently depend on `type` to be always present, but realized that this should gracefully fall back to the saved post's `type` property. Nice 👍

File: client/state/ui/editor/reducer.js [hunk 5]
```diff
@@ -21,8 +21,10 @@ import contactForm from './contact-form/reducer';
  */
 export function postId( state = null, action ) {
 	switch ( action.type ) {
-		case EDITOR_POST_ID_SET:
+		case EDITOR_START:
 			return action.postId;
+		case POST_SAVE_SUCCESS:
+			return state === action.postId ? action.savedPost.ID : state;
```
Comment (aduth) [inline client/state/ui/editor/reducer.js]: I guess we need this condition since other posts could potentially be saved in the background that aren't related to the one we're editing? I guess it's okay, since we're effectively ensuring that the ID is kept in sync after a save (especially for draft -> saved post).

File: client/state/ui/editor/actions.js [hunk 6]
```diff
@@ -20,13 +20,33 @@ export const MODAL_VIEW_STATS = {
  * Returns an action object to be used in signalling that the editor should
  * begin to edit the post with the specified post ID, or `null` as a new post.
  *
+ * @param  {Number}  siteId   Site ID
+ * @param  {?Number} postId   Post ID
+ * @param  {String}  postType Post Type
+ * @return {Object}           Action object
+ */
+export function startEditingPost( siteId, postId, postType ) {
```
Comment (aduth) [inline client/state/ui/editor/actions.js]: Do you think we should default `postType = 'post'`? Fine to not, though "post" is something of a default in WordPress (see [`wp_insert_post`](https://developer.wordpress.org/reference/functions/wp_insert_post/), [`WP_Query`](https://codex.wordpress.org/Class_Reference/WP_Query#Type_Parameters) `post_type` arguments).

File: client/post-editor/post-editor.jsx (hunk 3 above)
Comment (youknowriad) [inline client/post-editor/post-editor.jsx]: Yeah, I noticed these notices as well :). I'm starting to think, we should avoid using the middleware to automatically trigger notices on request success. I had to revert my own save settings global notices here https://github.com/Automattic/wp-calypso/pull/10278 to avoid double notices. I think we should rely on `componentWillReceiveProps` to trigger these notices.

File: client/state/posts/reducer.js (hunk 4 above)
Comment (youknowriad) [inline client/state/posts/reducer.js]: About that, I think we should move the `type` out of the `edits` state. I don't think it should be stored there since it's not something editable. I would store this on the `ui` subtree, the same way we store the current edited ID. Maybe another PR

File: client/state/ui/editor/actions.js (hunk 6 above)
Comment (youknowriad) [inline client/state/ui/editor/actions.js]: Yeah, I'm all for more consistency between APIs.

File: client/state/ui/editor/reducer.js (hunk 5 above)
Comment (youknowriad) [inline client/state/ui/editor/reducer.js]: Also, in theory, we could switch the editor's post before the SAVE_SUCCESS is dispatched. but very unlikely to happen.

File: client/post-editor/post-editor.jsx (hunk 3 above)
Comment (aduth) [inline client/post-editor/post-editor.jsx]: Alternatively we could create separate extended action creators for a more specific variations if we only want a notice to be shown in some cases, e.g. adding a meta flag that the middleware checks the presence of.

File: client/state/posts/reducer.js (hunk 4 above)
Comment (aduth) [inline client/state/posts/reducer.js]: It is an edit in that it needs to be sent with the initial payload for a new draft of non-post types. Though you're right it's not really editable for saved items and posts of type "post".

File: client/post-editor/controller.js [hunk 7]
```diff
@@ -177,6 +175,15 @@ module.exports = {
 		renderEditor( context, postType );
 	},
 
+	exitPost: function( context, next ) {
+		const postId = getPostID( context );
+		const siteId = getSelectedSiteId( context.store.getState() );
```
Comment (aduth) [inline client/post-editor/controller.js]: I'm seeing a few errors here on the Calypso errors dashboard:

>Cannot read property 'getState' of undefined

It's not clear to me when or why `context.store` would be ever be `undefined`.

Comment (youknowriad) [inline client/post-editor/controller.js]: Maybe it's because https://github.com/Automattic/wp-calypso/blob/master/client/state/initial-state.js#L90 is async. It seems that the context is initialized on the callback of this function https://github.com/Automattic/wp-calypso/blob/master/client/boot/index.js#L260

Is this the only place we have this kind of errors, seems surprising to me

Comment (aduth) [inline client/post-editor/controller.js]: @youknowriad Following the logic from boot, I don't think we render anything until the initial state has finished loading. It may have something to do with it being a `page.exit` handler, which we don't use in very many (any?) other parts of Calypso.
